src/Lecture2.hs

```diff
@@ -36,260 +42,145 @@ module Lecture2
-
-In programming we write expressions like "x + 1" or "y + x + 10".
-Such expressions can be represented in a more structured way (than a
-string) using the following recursive Algebraic Data Type:
--}
 data Expr
     = Lit Int
     | Var String
     | Add Expr Expr
     deriving (Show, Eq)
 
-{- Now, you can use this data type to describe such expressions:
-
-> x + 1
-Add (Var "x") (Lit 1)
-
-> y + x + 10
-Add (Var "y") (Add (Var "x") (Lit 10))
--}
-
-{- | We want to evaluate such expressions. We can associate a value
-with a variable using a list of pairs.
-
-You can use the @lookup@ function to search in this list by a variable name:
-
- * https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:lookup
--}
 type Variables = [(String, Int)]
 
-{- | Unfortunately, it's not guaranteed that variables in our @Expr@
-data type are present in the given list. So we're going to introduce a
-separate data for possible evaluation errors.
-
-Normally, this would be a sum type with several constructors
-describing all possible errors. But we have only one error in our
-evaluation process.
--}
 data EvalError
     = VariableNotFound String
     deriving (Show, Eq)
 
-{- | Having all this set up, we can finally implement an evaluation function.
-It returns either a successful evaluation result or an error.
--}
-eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
-
-{- | Compilers also perform optimizations! One of the most common
-optimizations is "Constant Folding". It performs arithmetic operations
-on all constants known during compile time. This way you can write
-more verbose and clear code that works as efficient as its shorter
-version.
 
-For example, if you have an expression:
-
-x + 10 + y + 15 + 20
-
-The result of constant folding can be:
-
-x + y + 45
-
-It also can be:
+eval :: Variables -> Expr -> Either EvalError Int
+eval var (Lit n) = Right n
+eval var (Var str) = case lookup str var of
+                        Nothing -> Left (VariableNotFound str)
+                        Just n  -> Right n
+eval var (Add e f) = combine ( if (isLeft $ eval var e) then Left $ fromLeft (VariableNotFound "s") (eval var e) else Right $ fromRight 1 (eval var e)) (if (isLeft $ eval var f) then Left $ fromLeft (VariableNotFound "s") (eval var f) else Right $ fromRight 1 (eval var f))
 
-x + 45 + y
+combine:: Either EvalError Int -> Either EvalError Int -> Either EvalError Int
+combine (Left (VariableNotFound str)) _ = Left (VariableNotFound str)
+combine _ (Left (VariableNotFound str)) = Left (VariableNotFound str)
+combine (Right a) (Right b) = Right (a+b)
 
-Write a function that takes and expression and performs "Constant
-Folding" optimization on the given expression.
--}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding expr = do
+  let (res, constSum) = constrExpr expr
+  case res of
+    Nothing -> Lit constSum
+    Just (Var str) -> Var str
+    Just exp -> Add exp (Lit constSum)
```

Comment:
It looks here that `constSum` will always be added as sum with `Add` even if it equals to zero 🤔 
Probably you can handle this extra case here 🙂